tests/common/hashkv_test.go

```diff
@@ -0,0 +1,248 @@
+			if tc.config.ClusterSize < 2 {
+				t.Skip("Skipping test for single-member cluster")
+			}
+			ctx, cancel := context.WithTimeout(t.Context(), 30*time.Second)
+			defer cancel()
+
+			clus := testRunner.NewCluster(ctx, t, config.WithClusterConfig(tc.config))
+			defer clus.Close()
+
+			cc := testutils.MustClient(clus.Client())
+			tombstoneRevs, latestRev := populateDataForHashKV(t, cc, []string{"key0"})
+			require.GreaterOrEqual(t, len(tombstoneRevs), 2)
+
+			firstCompactOnRev := tombstoneRevs[0]
+			t.Logf("COMPACT rev=%d", firstCompactOnRev)
+			_, err := cc.Compact(ctx, firstCompactOnRev, config.CompactOption{Physical: true})
+			require.NoError(t, err)
+
+			secondCompactOnRev := tombstoneRevs[1]
+			t.Logf("COMPACT rev=%d", secondCompactOnRev)
+			_, err = cc.Compact(ctx, secondCompactOnRev, config.CompactOption{Physical: true})
+			require.NoError(t, err)
+
+			for rev := secondCompactOnRev; rev <= latestRev; rev++ {
+				verifyConsistentHashKVAcrossAllMembers(t, cc, rev)
+			}
+		})
+	}
+}
+
+// TestVerifyHashKVAfterCompactOnLastTombstone tests that HashKV is consistent
+// across all members after a physical compaction on the last tombstone revision.
+func TestVerifyHashKVAfterCompactOnLastTombstone(t *testing.T) {
+	testRunner.BeforeTest(t)
+
+	for _, keys := range [][]string{
+		{"key0"},
+		{"key0", "key1"},
+	} {
+		t.Run(fmt.Sprintf("Keys=%v", keys), func(t *testing.T) {
+			for _, tc := range clusterTestCases() {
+				t.Run(tc.name, func(t *testing.T) {
+					if tc.config.ClusterSize < 2 {
+						t.Skip("Skipping test for single-member cluster")
+					}
+
+					ctx, cancel := context.WithTimeout(t.Context(), 10*time.Second)
+					defer cancel()
+
+					clus := testRunner.NewCluster(ctx, t, config.WithClusterConfig(tc.config))
+					defer clus.Close()
+
+					cc := testutils.MustClient(clus.Client())
+					tombstoneRevs, latestRev := populateDataForHashKV(t, cc, keys)
+					require.NotEmpty(t, tombstoneRevs)
+
+					compactOnRev := tombstoneRevs[len(tombstoneRevs)-1]
+					t.Logf("COMPACT rev=%d", compactOnRev)
+					_, err := cc.Compact(ctx, compactOnRev, config.CompactOption{Physical: true})
+					require.NoError(t, err)
+
+					for rev := compactOnRev; rev <= latestRev; rev++ {
+						verifyConsistentHashKVAcrossAllMembers(t, cc, rev)
+					}
+				})
+			}
+		})
+	}
+}
+
+// TestVerifyHashKVAfterCompactAndDefrag tests that HashKV is consistent
+// within a member before and after a physical compaction and defragmentation.
+func TestVerifyHashKVAfterCompactAndDefrag(t *testing.T) {
+	testRunner.BeforeTest(t)
+
+	for _, tc := range clusterTestCases() {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(t.Context(), 10*time.Second)
+			defer cancel()
+
+			clus := testRunner.NewCluster(ctx, t, config.WithClusterConfig(tc.config))
+			defer clus.Close()
+
+			cc := testutils.MustClient(clus.Client())
+			tombstoneRevs, _ := populateDataForHashKV(t, cc, []string{"key0"})
```

Comment:
In the previous integration test, the number of operations to populate sample data was more than 1,000.
https://github.com/etcd-io/etcd/blob/991ea7e0b4762303878f5b3c0cb4edc68c16bd39/server/storage/mvcc/testutil/hash.go#L31-L83


For now, we keep it at ~ 40 operations using `populateDataForHashKV`. 

This test validates an invariant rather than throughput, so higher operation counts don't improve the quality of the test.